rust-runtime/aws-smithy-observability/src/meter.rs

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Metrics are used to gain insight into the operational performance and health of a system in
+//! real time.
+
+use crate::attributes::{Attributes, Context};
+use crate::error::ObservabilityError;
+
+/// Provides named instances of [Meter].
+pub trait MeterProvider {
+    /// Get or create a named [Meter].
+    fn get_meter(&self, scope: &'static str, attributes: Option<&Attributes>) -> Box<dyn Meter>;
+
+    /// Optional method to flush the metrics pipeline, default is noop
+    fn flush(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+
+    /// Optional method to shutdown the metrics provider, default is noop
+    fn shutdown(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+}
+
+/// The entry point to creating instruments. A grouping of related metrics.
+pub trait Meter {
+    /// Create a new Gauge.
+    #[allow(clippy::type_complexity)]
+    fn create_gauge(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = f64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = f64>>;
+
+    /// Create a new [UpDownCounter].
+    fn create_up_down_counter(
+        &self,
+        name: String,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn UpDownCounter>;
+
+    /// Create a new AsyncUpDownCounter.
+    #[allow(clippy::type_complexity)]
+    fn create_async_up_down_counter(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = i64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = i64>>;
+
+    /// Create a new [MonotonicCounter].
+    fn create_monotonic_counter(
+        &self,
+        name: String,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn MonotonicCounter>;
+
+    /// Create a new AsyncMonotonicCounter.
+    #[allow(clippy::type_complexity)]
+    fn create_async_monotonic_counter(
+        &self,
+        name: String,
+        callback: Box<dyn Fn(&dyn AsyncMeasurement<Value = u64>) + Send + Sync>,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn AsyncMeasurement<Value = u64>>;
+
+    /// Create a new [Histogram].
+    fn create_histogram(
+        &self,
+        name: String,
+        units: Option<String>,
+        description: Option<String>,
+    ) -> Box<dyn Histogram>;
```

Comment:
We seem to be forcing `Box<dyn T>` everywhere and I'm wondering if this is the right tradeoff/choice. We probably want to be leveraging generics a bit more here.

Reply:
Moved to `Arc` in https://github.com/smithy-lang/smithy-rs/pull/3912/commits/4e11e53a2360c050313d71c067143d8db8ac873a